### PR TITLE
iterative gradient descent using theno

### PR DESCRIPTION
**Pull request by sadaf2605**
Implemented Iterative Gradient Descent [2D] using theno, tested using sklearn, plots look good.

### CODE REVIEW

**Comment by sadaf2605** [comment]
Opz, accidently I commented line 21.

**Comment by sadaf2605** [comment]
Did I violated PEP8 again? dammit! This time I was using sublime-text-2 and I will read the whole thing again!

**Comment by MartinThoma** [comment]
Copy your code to http://pep8online.com/ - then you will see that there are plenty of PEP8 violations.
I can recommend the "Python Checker" plugin for sublime text to check PEP8 automatically.

**Comment by sadaf2605** [comment]
I have uncommented the code! Can you give it a look? It works fine in my end...

figure_1

**Comment by MartinThoma** [comment]
Could you please explain your code? It seems not to implement gradient descent. For example, where do you calculate the gradient?

The code seems to make only a few matrix operations (additions and dot products). What are your neurons? What represents hidden layers?

By the way, have you seen https://github.com/MartinThoma/nntoolkit/blob/master/nntoolkit/train.py

**Comment by MartinThoma** [comment]
The image and the code looks rather as if you implemented linear regression.

**Comment by sadaf2605** [comment]
OOpz, in issue 4 (https://github.com/MartinThoma/nntoolkit/issues/4) you were talking about "batch gradient descent" so I thaught you were talking about http://en.wikipedia.org/wiki/Stochastic_gradient_descent. Probably I have mistaken something then :/

**Comment by MartinThoma** [comment]
There are three variants of gradient descent:

* Batch gradient descent: Takes all training examples into consideration, then improves the weights
* Mini-batch gradient descent: Takes a couple (typically 256) training examples into consideration, then improves the weights
* Stochastic gradient descent: Improves the weights after every single training example.

I want to implement mini-batch gradient descent, as you can change a parameter and get both other variants.

You implementations seems not to be able to be generizable to fit higher-order function (higher than linear) to data. Am I wrong?
Neural networks are typically very high dimensional functions (e.g. R^160 -> R^369).

**Comment by sadaf2605** [comment]
Alright leave it here. I think I messed up :P

**Comment by sadaf2605** [comment]
I mean it is gradient descent regression, you were looking for classification [for nn]. Alright, I also need to read about nn, I will get back after that.
